### PR TITLE
[libc++] Fix constexpr initialization of std::array<T, 0>

### LINKED ISSUE #74375: [libc++] Initialization of std::array of size 0 in constexpr contexts
```c++
constexpr std::array<int, 0> a;
```
This fails to compile with libc++ (https://godbolt.org/z/drcGsW84x), since its zero-length `std::array` specialization has some dummy data (see #34839) that is left uninitialized.

This is quite unfortunate for a case like
```c++
template <int N>
constexpr auto getOnes() {
    std::array<int, N> a;
    a.fill(1);
    return a;
}
```
which will fail to compile only when used in a `constexpr` context with `N = 0` (https://godbolt.org/z/jxMGx1941). Using `std::array<int, N> a{};` fixes that, but it may also hide programming errors (i.e. missed elements) in the following actually intended initialization code for `N > 0` that would otherwise be diagnosed.

Microsoft's STL was in a very similar situation and fixed this just recently (see https://github.com/microsoft/STL/pull/3863), so I was hoping that maybe libc++ might want to follow. Seems like for libstdc++ this was never a problem since their `std::array<T, 0>` is actually empty.

### PR DESCRIPTION
**Pull request by ldionne**
This patch fixes constexpr default initialization of empty arrays and improves the tests accordingly.

Fixes #74375

### CODE REVIEW

File: libcxx/include/array
Comment (mordante) [inline libcxx/include/array:284]:
```suggestion
    using _EmptyType = __maybe_const<_Tp>::value, _EmptyAggregate>;
```
// #include <__type_traits/maybe_const.h> is needed too
The name was not great before, but now it looks really weird since there is no connection with `char` at all.

File: libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp
Comment (mordante) [inline libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp:104]: Maybe remain `main` to `test` since it's a compile only test.

**Review by mordante — APPROVED** [review]
LGTM modulo some nits.

**Review by philnik777 — CHANGES_REQUESTED** [review]
I'd really like to see a test that makes sure the new struct isn't considered empty and there are no padding bytes.

File: libcxx/include/array
Comment (ldionne) [reply libcxx/include/array:284]: I agree with renaming to `_EmptyType`, that was an oversight. However, I'd like to keep the `__maybe_const` out of this change, since it's really unrelated.

File: libcxx/include/array
Comment (philnik777) [inline libcxx/include/array:283]: We have `__empty` and I don't think we have to have this be specific to the instantiation, so maybe we want to use that instead?

File: libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp
Comment (philnik777) [inline libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp:184]: I like this!

File: libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp
Comment (philnik777) [inline libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp:42]: These look to me like they're all implementation-dependent. Am I missing some requirement, or should we make them libc++-specific?

Comment (ldionne) [reply libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp:42]: cppreference says:

> This container is an aggregate type with the same semantics as a struct holding a C-style array `T[N]` as its only non-static data member.

I think it's reasonable (and helpful for other implementations) to have these checks enabled for all implementations.

The tests for the 0-sized specialization could potentially be made libc++ only, but then again I think this will help enforce that all implementations provide compatible implementations.

Comment (ldionne) [reply libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp:42]: I see, I thought implementations agreed on this a bit more than they actually do.

**Comment by EricWF** [comment]
Hmm, So I see we already migrated away from  not returning `nullptr` from `array<T, 0>`. My understanding was that was explicitly prohibited by the standard when it says 

[2](https://eel.is/c++draft/sequences#array.zero-2)
[#](http://github.com/Eelis/draft/tree/180690ab2654988b5d40bfe2f50b32737f02f20b/source/containers.tex#L6391)
In the case that N == 0, begin() == end() == unique value[.](https://eel.is/c++draft/sequences#array.zero-2.sentence-1) The return value of data() is unspecified[.](https://eel.is/c++draft/sequences#array.zero-2.sentence-2)

The "unique" bit specifically saying that the two arrays can't both return nullptr.

I think we may have strayed from the right implementation here, and done it a while ago. I would like to consider this more.

File: libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp
Comment (EricWF) [inline libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp:31]: I really hate logic in tests. The point where we have this many conditionals is the point where we should probably make the tests less generic and just spam the test out multiple times.

File: libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp
Comment (EricWF) [inline libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp:21]: Please make these separate tests rather than trying to reuse the instantiation. It's just clearer for readers.

File: libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp
Comment (EricWF) [inline libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp:14]: Only partially related but...

Any chance we've changed the triviality of any of the special members with this change?

**Review by EricWF — APPROVED** [review]
OK, I've done my due diligence offline, and this seems like a not incorrect fix.

Though as a general note, I think our tests have become too clever.

**Comment by ldionne** [comment]
> Hmm, So I see we already migrated away from not returning `nullptr` from `array<T, 0>`. My understanding was that was explicitly prohibited by the standard when it says
> 
> [2](https://eel.is/c++draft/sequences#array.zero-2) [#](https://github.com/Eelis/draft/tree/180690ab2654988b5d40bfe2f50b32737f02f20b/source/containers.tex#L6391) In the case that N == 0, begin() == end() == unique value[.](https://eel.is/c++draft/sequences#array.zero-2.sentence-1) The return value of data() is unspecified[.](https://eel.is/c++draft/sequences#array.zero-2.sentence-2)
> 
> The "unique" bit specifically saying that the two arrays can't both return nullptr.
> 
> I think we may have strayed from the right implementation here, and done it a while ago. I would like to consider this more.

Just to be clear though, this patch doesn't affect whether we return `nullptr` or not from `begin()`, nor our ability to do so. It only fixes cases where we should be able to initialize an empty array inside `constexpr` and we currently can't, without breaking the ABI. If you wanted to return `nullptr` from `begin()` (which I think you can't without contradicting the standard), you could in theory do it orthogonally to this patch.

File: libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp
Comment (ldionne) [reply libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp:31]: I know -- I started like that and it was unreadable since I could hardly reuse anything at all, and I'm spamming exactly the same test over and over again for a bunch of different types. I dislike logic in tests, but I think this strikes the right balance after spending a bunch of time investigating different approaches.

File: libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp
Comment (ldionne) [reply libcxx/test/std/containers/sequences/array/array.cons/initialization.pass.cpp:21]: The only reason this was multiple structs previously is basically that I couldn't run the default initialization test in constexpr, so I needed another named struct for it. The way this is written right now follows our usual pattern of having something like:

```
constexpr bool test() {
  // test case 1
  {
    ...
  }
  // test case 2
  {
    ...
  }
  // test case 3
  {
    ...
  }

  return true;
}

int main() {
  test();
  static_assert(test());
}
```

File: libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp
Comment (ldionne) [reply libcxx/test/std/containers/sequences/array/size_and_alignment.compile.pass.cpp:14]: That's a great observation! This is tested in https://github.com/llvm/llvm-project/blob/main/libcxx/test/libcxx/containers/sequences/array/triviality.pass.cpp and I think this already has the appropriate coverage, but I am happy to add to it if you think we're missing something there.

**Comment by rnicholl-google** [comment]
> > Hmm, So I see we already migrated away from not returning `nullptr` from `array<T, 0>`. My understanding was that was explicitly prohibited by the standard when it says
> > [2](https://eel.is/c++draft/sequences#array.zero-2) [#](https://github.com/Eelis/draft/tree/180690ab2654988b5d40bfe2f50b32737f02f20b/source/containers.tex#L6391) In the case that N == 0, begin() == end() == unique value[.](https://eel.is/c++draft/sequences#array.zero-2.sentence-1) The return value of data() is unspecified[.](https://eel.is/c++draft/sequences#array.zero-2.sentence-2)
> > The "unique" bit specifically saying that the two arrays can't both return nullptr.
> > I think we may have strayed from the right implementation here, and done it a while ago. I would like to consider this more.
> 
> Just to be clear though, this patch doesn't affect whether we return `nullptr` or not from `begin()`, nor our ability to do so. It only fixes cases where we should be able to initialize an empty array inside `constexpr` and we currently can't, without breaking the ABI. If you wanted to return `nullptr` from `begin()` (which I think you can't without contradicting the standard), you could in theory do it orthogonally to this patch.
> 
> Edit: It looks like we raced on this! Edit2: My comment is actually completely wrong. We are returning `nullptr` from `begin()` since [7265ff9](https://github.com/llvm/llvm-project/commit/7265ff928a974a844b6301c139cbb0f957532da9). And yes I agree with you that this is technically not standards-conforming, however that commit explains why all other implementations are also not standards-conforming.

I may have found a possible way to do a standard compliant implementation, so don't merge this yet, I'll try to have a patch within the week. I'm not sure if it'll work fully yet and pass all tests but I found a hack with unions that might make the non-nullptr version implementable.

**Comment by rnicholl-google** [comment]
Update: My fix doesn't work because of a different clang bug, so go ahead and merge this if it works, I think I can separately merge the other one if the other bug is fixed.

**Comment by ldionne** [comment]
I will merge this since it passes CI and looks like folks are satisfied. @EricWF if you want to talk about how I wrote the tests, I'm happy to go over it and change them in a future patch once we agree.

**Pull request merged**
